features/237-cms-validation/api.md

```diff
@@ -0,0 +1,146 @@
+# CMS Validation
+
+## Requirements
+
+- Prevent cms user from entering invalid data.
+- Allow developer to use common and customisable validation in cms config on field level and form level.
+- Display error message, which can be defined in config, for invalid data.
+
+## Common Use Cases
+
+- Non empty text input.
+- Text input with max length.
+- Text input with certain pattern, e.g. credit card number.
+- Number or datetime input value within range.
+- One field cannot be empty if another field is not empty.
+
+## Overview
+
+- Validation is a process that evaluate data with an expression, the result is either success or failure.
+- Validation can be defined in record edit page and new page in cms config.
+- Validation happens after use click save button and before all server api calls are fired.
+- If any validation result is failure, error message should be displayed and no server api calls should be fired.
+- Error message of field validation appears next / below the field, while that of form validation appears below the form.
+- Field validation is executed before form validation.
+
+## Configuration
+
+- One validation item is a dictionary which contains the validation configuration (depends on where it is placed).
+  - Each validation optionally contains a `message` field, which would be the error message if that validation item cannot pass.
+  - Otherwise, a default error message would be shown.
+- `validation` may be a validation item or contains a list of validation items.
+
+```
+new:
+  # page validation
+  validation:
+    - expression: xxx
+    - expression: xxx
+  fields:
+    - name: xxx
+      type: xxx
+      # field validation
+      validation:
+        pattern: /^xxx$/
+        message:
+```
+
+## Expression
+
+- `expression` is the base type of validation, supported by all validation.
+- An expression is a string, which can be evaluated to give boolean value.
+- All other types of validation are predifined expression, i.e. they are all compiled to expression at runtime.
+- Expression evaluation would be based on this library https://github.com/joewalnes/filtrex, while CMS would provide extra predefined functions for common use case.
+
+### Functions
+
+- Passing wrong data type to the function would throw error in validation.
+
+### Expression context: `value`
+
+Developer uses the variable `value` in the expression to get the value from the field or form.
+
+- Field validation
+  - `value`: value of the field
+  - e.g. `value in ('abc', 'def')`
+- Form validation
+  - `value`: value of the form
+  - `get(value: formValue, fieldName: string)`
+  - e.g. `get(value, "full_name") != get(value, "nick_name")`
+
+#### Predefined functions
+
+- String
+  - `length(value: string)`
+  - `upper(value: string)`
+  - `lower(value: string)`
+  - `substring(value: string, from: number, to: number)`
+- Number
+- Datetime
```

Comment:
should we have some function for date range operation? Like "with in 7 days" or "within the same month" / "within the same week" etc?

Reply:
So there should be some functions like `getMonth(value)` or `getWeekOfYear(value)`?

Reply:
> So there should be some functions like getMonth(value) or getWeekOfYear(value)?

Agree, I will add those functions, and with a function `now()`, probably would do what @chpapa mentioned.

Reply:
I'm not familiar with the expression language, so someone should try to look at it and consider if it make sense to make `getMonth(value)`/`getWeekOfYear(value)`, or something like `daysAfter(date, n)`.

My problem with `getMonth` or `getDay` etc, is usually it is not straightforward, for example, if you get the day is 28, and month is 5, and wanna make it 5 days later, you can never simply add 5 to day (28+5=33, and obviously there is no May 33rd)

Reply:
maybe we can handle this by the timestamp function?